File: lab3_parallelization/scripts/parallel_example_with_args.py
import pandas as pd
import numpy as np
import pickle as pkl
import sys
import os
from os.path import join as oj
import time
import argparse
from joblib import Parallel, delayed

sys.path.append(".")
from python.fit import fit_rf_loo, fit_knn_loo

if __name__ == '__main__':

    # read in command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("--array_id", type=int, default=1, help="Array ID")
    args = parser.parse_args()

    # select the method name based on the array_id
    method_names = ["rf", "knn"]
    method_name = method_names[args.array_id]

    # Helper variables
    DATA_PATH = oj("data")
    RESULTS_PATH = oj("results")
    n_cores = int(os.getenv("NSLOTS"))
    print("Number of cores available:", n_cores)
    n = n_cores * 10  # Do leave-one-out for the first n samples

    X = pd.read_csv(oj(DATA_PATH, "X_tcga_cleaned.csv")).values  # Convert to NumPy array
    y = pd.read_csv(oj(DATA_PATH, "Y_tcga.csv")).iloc[:, 0].values  # Convert to NumPy array

    if method_name == "rf":
        fit_loo = fit_rf_loo
    elif method_name == "knn":
        fit_loo = fit_knn_loo
    else:
        raise ValueError("Unknown method name")

    start_time = time.time()
    preds = Parallel(n_jobs=n_cores)(delayed(fit_loo)(i, X, y) for i in range(n))
    end_time = time.time()
    execution_time = end_time - start_time
    print("Execution time with joblib parallelization:", execution_time)

    os.makedirs(RESULTS_PATH, exist_ok=True)
    with open(oj(RESULTS_PATH, f"preds_{method_name}.pkl"), "wb") as f:
        pkl.dump(preds, f)

File: lab3_parallelization/scripts/parallel_example_with_args.R
# to submit job: qsub -N parallel_example_with_args submit_r_array_job.sh scripts/parallel_example_with_args
library(future)
library(furrr)
library(optparse)

# define the list of command line options
option_list <- list(
  make_option(
    "--array_id", type = "integer", default = 1, 
    help = "array id", metavar = "integer"
  )
)

# parse the command line options
opt_parser <- OptionParser(option_list = option_list)
opt <- parse_args(opt_parser)

# translate option into selected method name
method_names <- c("rf", "knn")
method_name <- method_names[opt$array_id]
print(method_name)

# load in helper functions
source(here::here("R", "fit.R"))

# set seed
set.seed(242)

# set up parallel processing
n_cores <- as.integer(Sys.getenv("NSLOTS"))
print(paste("Using the following number of cores:", n_cores))
plan(multicore, workers = n_cores)

# helper variables
n <- n_cores * 10  # do leave-one-out for this many number of samples
DATA_DIR <- here::here("data")
RESULTS_DIR <- here::here("results")

# load in TCGA breast cancer data
X <- data.table::fread(
  here::here(DATA_DIR, "X_tcga_cleaned.csv"),
  data.table = FALSE
)
y <- data.table::fread(
  here::here(DATA_DIR, "Y_tcga.csv"),
  data.table = FALSE
)[, 1] |>
  as.factor()

if (method_name == "rf") {
  fit_loo <- fit_rf_loo
} else if (method_name == "knn") {
  fit_loo <- fit_knn_loo
} else {
  stop("Invalid method name")
}

start_time <- Sys.time()
preds_furrr <- furrr::future_map(
  1:n,
  function(i) {
    fit_loo(i, X, y)
  },
  .options = furrr::furrr_options(seed = TRUE)
)
end_time <- Sys.time()
execution_time <- end_time - start_time
print(execution_time)

if (!dir.exists(RESULTS_DIR)) {
  dir.create(RESULTS_DIR)
}
saveRDS(
  preds_furrr, 
  file = here::here(RESULTS_DIR, sprintf("preds_%s.rds", method_name))
)